### PR TITLE
Prototype creating a WinGet Configuration DSC file in PowerShell

### PR DESCRIPTION
**Pull request by KevinLaMS**
- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [x] Are you working against an Issue?

-----
Creating DSC files can be a challenge.
This script tries to make it easier by leveraging winget search to locate the packages, and then write the file out as YAML.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/winget-create/pull/462)

### CODE REVIEW

**Comment by KevinLaMS** [comment]
#458 @Trenly your feedback would be appreciated too.

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (Trenly) [inline src/prototype/WingetCreateMakeDSC.ps1:10]: The Install-Module calls should probably be inside Try-Catch that capture any installation errors and re-throw them as a missing dependency error since Install-Module is not always successful

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (Trenly) [inline src/prototype/WingetCreateMakeDSC.ps1:5]: Does this still require PS Core (v7)? If so, should probably add a #Requires statement to ensure the script can only run in Powershell 7

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (Trenly) [inline src/prototype/WingetCreateMakeDSC.ps1:49]: Shouldn’t the source here come from the package the user selected? In theory it would be possible for them to have selected an msstore package, unless you update the initial search to only query the winget source

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (Trenly) [inline src/prototype/WingetCreateMakeDSC.ps1:62]: I would suggest to explicitly set the encoding on the output file to UTF-8, just to be sure that the output file is always consistent with the encoding winget uses

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (mdanish-kh) [inline src/prototype/WingetCreateMakeDSC.ps1:37]: `$selection -gt $findResult.count` will incorrectly allow one greater than the last value to be a valid selection with the current logic. I would suggest keeping this check as is but do the following:
- initiating $index from 1 (since it's a user-facing number) 
- change ($selection -lt 0) to `($selection -lt 1)`
- change `$selectedPackage = $findResult[$selection]` to be `$selectedPackage = $findResult[$selection - 1]`

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (mdanish-kh) [inline src/prototype/WingetCreateMakeDSC.ps1:63]: Before we print this out, would it be a good idea to run `winget configure validate` on the generated file?

We can add a check if user does not have `winget` on their machine, or even a previous version of winget that does not have `winget configure`

```
if (Get-Command 'winget configure' -ErrorAction SilentlyContinue) { winget configure validate $filePath }
```

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (Trenly) [inline src/prototype/WingetCreateMakeDSC.ps1:30]: This can actually be achieved using the pipeline, and might be cleaner
```powershell
$findResult | ForEach-Object {$i=1} {Add-Member -InputObject $_ -NotePropertyName Index -NotePropertyValue $i; $i++ }
$findResult | Select-Object -Property Index,Name,Id,Version
```

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (Trenly) [reply src/prototype/WingetCreateMakeDSC.ps1:37]: With my above suggestion adding the index as a property to each package, this would also be simplified
* `if $selection -notin $findResult.Index` can be used to check if the index is in range
* `$selectedPackage = $findResult.Where({$_.Index -eq $selection})` can be used to set the selected package

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (KevinLaMS) [reply src/prototype/WingetCreateMakeDSC.ps1]: Great idea. Completed.

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (KevinLaMS) [reply src/prototype/WingetCreateMakeDSC.ps1]: I went with @mdanish-kh  proposal. I played around with Trenly's but I fell over.  :)

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (KevinLaMS) [reply src/prototype/WingetCreateMakeDSC.ps1:111]: Done

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (KevinLaMS) [reply src/prototype/WingetCreateMakeDSC.ps1]: I wanted to get an update out, so I skipped this one for now.

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (Trenly) [reply src/prototype/WingetCreateMakeDSC.ps1]: I'll take a look again and might make a PR to your fork if it would be beneficial

**Comment by KevinLaMS** [comment]
@Trenly @mdanish-kh 
1) Thanks for the feedback. I have tried to implement.
2) I resolved the threads, but in retrospect, I probably should have let you do that.  Sorry.
3) Menace me some more, if I messed up your recommendations.

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (KevinLaMS) [reply src/prototype/WingetCreateMakeDSC.ps1]: I tried to address this.

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (mdanish-kh) [inline src/prototype/WingetCreateMakeDSC.ps1:83]:
```suggestion
    [void]$finalPackages.Add($unit)
```

nit: Instead of assigning to an unused variable to discard the index printed to stdout, you can cast the result to void.

**Review by mdanish-kh — COMMENTED** [review]
@KevinLaMS 

> I resolved the threads, but in retrospect, I probably should have let you do that. Sorry.

GitHub allows only the PR author or users with write access to resolve threads, so you did the right thing ☺️
   
> Menace me some more, if I messed up your recommendations.

Did some more menacing😄Most are nits that we can choose to leave and iterate upon them after merging in this PR

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (mdanish-kh) [reply src/prototype/WingetCreateMakeDSC.ps1:5]: Alternatively, we can just add 
```
#Requires -Version 7.0
```

to the top of the script and let pwsh automatically handle the error message. (though we wouldn't get our custom message that way)

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (Trenly) [reply src/prototype/WingetCreateMakeDSC.ps1:76]: I've got an even better solution; Will raise a PR against this branch

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (Trenly) [reply src/prototype/WingetCreateMakeDSC.ps1:96]: I believe there is work going on over at cli to allow double clicking .winget files for automatic install

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (Trenly) [reply src/prototype/WingetCreateMakeDSC.ps1:98]: Fixing as part of separate pr into this branch

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (Trenly) [reply src/prototype/WingetCreateMakeDSC.ps1:37]: Fixing as part of separate pr into this branch

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (Trenly) [reply src/prototype/WingetCreateMakeDSC.ps1:7]: Fixing as part of separate pr into this branch

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (mdanish-kh) [reply src/prototype/WingetCreateMakeDSC.ps1:96]: Ah, I see. I thought that was only for `.wingetdev` files. Read the [PR](https://github.com/microsoft/winget-cli/pull/3860) description again and it seems `.winget` is also supported

File: src/prototype/WingetCreateMakeDSC.ps1
Comment (Trenly) [reply src/prototype/WingetCreateMakeDSC.ps1:32]: Will address as part of my PR

**Review by ryfu-msft — APPROVED** [review]

**Pull request merged**
